### ISSUE TITLE
Automatic differentiation?

### ISSUE THREAD

**Issue by swertz** [zfit/zfit#161]
Apologies if the answer to my question is obvious, but do the various minimizers included actually use TensorFlow's autodiff capabilities, or is TF only used to build to model and evaluate the likelihood?

What about analytically computing the error matrix?

**Comment by mayou36** [comment]
It's a good question, thanks! zfit uses the full capabilities of TensorFlow including the automatic gradient. Minuit, as well as most other minimizer packages (scipy, ...), provide the possibility to give the gradient, this is done in zfit by using the automatic (~analytic) gradient provided by TensorFlow. You can obtain the (smartly cached) gradient of any loss by calling
```
loss.gradients(params)  # with a list of the parameters
```

The speedup of the minimization process due to the automatic gradient depends on the fitting function, in our experience this effect varies between 0-40% (mostly for more complex pdfs).

Computing the error matrix (hessian approximation of the error) is _currently_ not done using this functionality. Two reasons:
 - the computation of it numerically is (usually) not a bottleneck (the fit takes longer). Therefore, it is not an immediate priority to implement that since the focus is to create a stable, well designed library. And there are dozens of things that can be filled, e.g. more pdfs, losses etcetcetc. But since this implementations (including the calculation of the error matrix) are "rather simple" (contributions welcome!) and small, they are more left to the community (or for the core team, but this takes more time).
 - higher order derivatives in TF 1.x (with controlflow v1) are not really well supported. With v2, they are (coming with TF 2.0, expected in ~3-4 months). So no need to rush anyway.

You mentioned the unbinned case with the normalization: this is not a problem. Or let's say, it's independent of whether you use the automatic gradient or do a numerical calculation of the gradient.

Let us know in case you are interest in contributing (or need something for your usecase, which often goes hand-in-hand), have a specific question about a fit you try to do or anything else.

**Comment by swertz** [comment]
Thanks a lot for clarifying! I've run into cases before where the Hessian returned by Minuit was problematic (not positive definite) at the minimum because of numerical issues (probably because of large correlations between some parameters), not because the likelihood was non parabolic: in those cases an analytical computation should really help.

For now I'll play around some more with the library to see if it fits my ultimate needs or if something's missing for my use case (I don't exclude contributing at some point but cannot make promises)...

**Comment by mayou36** [comment]
yes, that should help indeed.

That's good, if you are unsure whether your specific fit works and want to know if it's capable, feel free to explain the idea and we can tell directly if it works.

(_N.B: right now, binned fits are not yet properly implemented, they are though a few lines of code to do by hand. It is currently the last major piece under development in zfit_)

Feel free to close

**Comment by swertz** [comment]
The goal is a multi-parameter template (i.e. binned) fit of several independent templates, with shape uncertainties on the templates, parameterised by nuisance parameters, that are inter-/extrapolated from the nominal and up/down variations of the templates, plus some Gaussian constraint terms. The nuisance parameters should be profiled when extracting the Hessian.

Is that realistic at this point?

**Comment by mayou36** [comment]
tl;dr: "no". 

Let me disentangle that:
 - binned pdfs and binned fits are _currently_ not implemented, although you can implement them by yourself (as a few have done already). But some TF knowledge and fun for that kind of thing should be a prerequisite
 - _any_ kind of parametrization is in zfit possible, so there is surely no limit on that. Build any pdf with parameters and they will be taken correctly into account.
 - Gaussian constraints (or even arbitrary) are well integrated.
 - profiling (currently with minos) is in zfit. Or you could even use the statistics library [lauztat](https://github.com/marinang/lauztat), which is independently built on top of zfit and offers such features (CLs etc)

So rather not. BUT: this is what we're fully working on and other have shown that it is fully possible (see also e.g. [this meeting](https://indico.cern.ch/event/827035/)) to do in TF.

I expect this feature to be implemented during September.

**Comment by swertz** [comment]
Thanks a lot for the feedback! It's certainly a very interesting and useful effort, I'll give it a try.
